### PR TITLE
Listview batch publishing on multi language site, publishes all variants #13755 - Fix

### PR DESCRIPTION
**Pull request by mennomout**
### Prerequisites

Link to issue: https://github.com/umbraco/Umbraco-CMS/issues/13755

### Description
Modified `PostPublishById` to handle content that is culture variant. Added a new model for the endpoint and modified the JavaScript function that calls said endpoint.

I've tested these changes by creating a new doctype with list view enabled. I've allowed another doctype as it's child. Both doctypes are culture variant. After these changes you can publish one culture from the list view and the other culture will remain unpublished. **I'd like someone else to test these changes as well.** There were a few occasions where it looked like it wasn't working but I wasn't really able to replicate that. 

See gif below for the result:
listviewsave

### CODE REVIEW

**Comment by mikecp** [comment]
Hi @mennomout ,
Thanks a lot for this PR, it's gonna fix an annoying problem 👍

I had a quick look at it and I would suggest to create a separate controller method for this scenario, and then to handle in the Javascript part which one will be called, depending on whether or not you have cultures specified: no culture => call as now, specific culture => new method.
The main reason is that I think that changing the signature of the controller will be seen as a breaking change, which we cannot merge in minor versions, and also this will allow you to get rid of that big if/else block 😁
So, in other words, doing the if/else at the Javascript side might make things a little less complex.

I was also wondering how the "cultures" param was sent through to the Javascript method since you don't seem to be modifying the caller(s) to that function, but this is probably because I did not actually run you solution locally yet 😅

I'll make sure to check in details later, but I thought it was already worth mentioning those two points.

Cheers!

**Comment by mennomout** [comment]
@mikecp While working on implementing your suggestions I've come across another bug related to the list view publishing action. When there is a content item with more then one culture, you are allowed to press the publish button (publish from list view) without selecting a culture. Selecting a culture and deselecting it makes it behave like it should (disables the publish button because it doesn't know what culture you want to publish to). Then if you do press publish without selecting a culture it publishes all cultures but shows up in the list view as: Not created.
noValidationOnCulturePublishButton

**Comment by mennomout** [comment]
That last comment was actually cause because not culture was selected in the contente tree:
image

I think it's because I was reconfiguring the website for the test scenario and should have refreshed the page or something along those lines.

**Comment by mennomout** [comment]
@mikecp I've made some changes based on your suggestions. Please let me know if there is anything else I can do to improve the PR.

File: src/Umbraco.Web.BackOffice/Controllers/ContentController.cs [hunk 1]
```diff
@@ -1960,6 +1960,58 @@ public IActionResult PostPublishById(int id)
...
+    public IActionResult PostPublishByIdAndCulture(PublishContent model)
+    {
+        IContent? foundContent = GetObjectFromRequest(() => _contentService.GetById(model.Id));
+
+        if (foundContent == null)
+        {
+            return HandleContentNotFound(model.Id);
+        }
+
+        var languageCount = _allLangs.Value.Count();
```
Comment (mikecp) [inline src/Umbraco.Web.BackOffice/Controllers/ContentController.cs:1982]: I'm wondering if this would not be better on top, for performance reasons: we also call `GetObjectFromRequest` in the `PostPublishById` method, which means that in the case that we end up calling `PostPublishById`, in the current situation we'll have two calls to `GetObjectFromRequest`, meaning two potential trips to the database to load the data.

If we move this part on top, then we'll have only one call in all situations I think.

**Review by mikecp — CHANGES_REQUESTED** [review]

**Comment by mikecp** [comment]
Hi @mennomout ,

Thanks for the update, it looks less complex in this new version, and it is also non breaking 😉 👍

It all works fine but I left a few suggestions that might improve performances a bit, let me know if they seem OK to you.

Cheers!

File: src/Umbraco.Web.BackOffice/Controllers/ContentController.cs (hunk 1 above)
Comment (mennomout) [reply src/Umbraco.Web.BackOffice/Controllers/ContentController.cs:1982]: This makes sense indeed. I've moved the check for publishing all cultures to the top of the method.

File: src/Umbraco.Web.BackOffice/Controllers/ContentController.cs [hunk 2]
```diff
@@ -1960,6 +1960,58 @@ public IActionResult PostPublishById(int id)
...
+        }
+
+
+        if (results.Any(x => x.Value.Success == false))
+        {
+            var notificationModel = new SimpleNotificationModel();
+
+            foreach (var culture in results)
+            {
+                if (culture.Value.Success == false)
```
[the remark this answers is not in the document]
Comment (mennomout) [reply src/Umbraco.Web.BackOffice/Controllers/ContentController.cs:2003]: What I've read about this is that using LINQ usually adds some overhead instead of improving speed because it uses loops internally as well. I've done some testing in a separate environment but my experience with benchmarking is not sufficient to draw any conclusions from these tests. However I do think using the Where query here improves the readability of the code slightly and therefore makes sense to add.

File: src/Umbraco.Web.UI.Client/src/common/resources/content.resource.js [hunk 3]
```diff
@@ -1089,20 +1089,27 @@ function contentResource($q, $http, umbDataFormatter, umbRequestHelper) {
          * @returns {Promise} resourcePromise object containing the published content item.
          *
          */
-        publishById: function (id) {
-
+        publishById: function (id, cultures) {
```
[the remark this answers is not in the document]
Comment (mennomout) [reply src/Umbraco.Web.UI.Client/src/common/resources/content.resource.js:1092]: Added the extra parameter.

**Comment by mennomout** [comment]
@mikecp I've implemented your suggested changes. Thank you again for the feedback!

File: src/Umbraco.Web.BackOffice/Controllers/ContentController.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikecp) [reply src/Umbraco.Web.BackOffice/Controllers/ContentController.cs]: Yes indeed it's not always easy to determine this kind of metrics.
But in this case the list will be relatively limited, so performance wise we should not see a big difference and, as you said, it improves readability.

**Review by mikecp — APPROVED** [review]

**Comment by mikecp** [comment]
Thanks a lot @mennomout for the adaptations!
This is a great fix that solves a bug I'm sure quite a lot of people were impacted by 👍

And congratulations on your first contribution to the Umbraco CMS 🎉
If you provide us with your account name on Our, @nul800sebastiaan will be very happy to assign our shiny contributor's badge to you 🥇

And now we'll be looking forward to your next PR"s 😁

Cheers!

**Comment by mennomout** [comment]
@mikecp Thank you for the kind words and assistance with this PR! 

@nul800sebastiaan My OurUmbraco profile is: https://our.umbraco.com/members/mennomout/

**Comment by cornehoskam** [comment]
Wonderful job @mennomout! Glad to see your first contribution merged into the source-code!

**Comment by mennomout** [comment]
Thank you both @nul800sebastiaan and @cornehoskam!

**Comment by nul800sebastiaan** [comment]
Cherry picked for 10.5 in 273e7708e55214fc23c9b9b779007bd287c8b2ef
